### PR TITLE
[fix] Protect against regex expression injection

### PR DESCRIPTION
**Pull request by merlimat**
### Motivation

Fix "Regular expression injection" issues reported by CodeQL

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (10MB)*
  - *Extended integration test for recovery after broker failure*

### Does this pull request potentially affect one of the following parts:

*If the box was checked, please highlight the changes*

- [ ] Dependencies (add or upgrade a dependency)
- [ ] The public API
- [ ] The schema
- [ ] The default values of configurations
- [ ] The threading model
- [ ] The binary protocol
- [ ] The REST endpoints
- [ ] The admin CLI options
- [ ] The metrics
- [ ] Anything that affects deployment

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete` 

### Matching PR in forked repository

PR in forked repository:

### CODE REVIEW

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java [hunk 1]
```diff
@@ -3040,7 +3040,8 @@ protected void handleCommandWatchTopicList(CommandWatchTopicList commandWatchTop
         final NamespaceName namespaceName = NamespaceName.get(commandWatchTopicList.getNamespace());
 
         Pattern topicsPattern = Pattern.compile(commandWatchTopicList.hasTopicsPattern()
-                ? commandWatchTopicList.getTopicsPattern() : TopicList.ALL_TOPICS_PATTERN);
+                ? Pattern.quote(commandWatchTopicList.getTopicsPattern())
```
Comment (lhotari) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java:3043]: Isn't `Pattern.quote` going to disable any wildcard matches and regex behavior?

**Comment by lhotari** [comment]
I guess this is about attempting to protect Pulsar Broker from [ReDoS attacks](https://owasp.org/www-community/attacks/Regular_expression_Denial_of_Service_-_ReDoS)?

There's probably not much that can be done when `java.util.regex.Pattern` is used and the user can pass the regex pattern.
Some regex libraries seem to have a backtracking limit to prevent such attacks or unintentional mistakes in regular expressions which cause [catastrophic backtracking](https://www.regular-expressions.info/catastrophic.html).

**Comment by lhotari** [comment]
ChatGPT is offering https://github.com/google/re2j as a possible solution. :)

**Comment by lhotari** [comment]
Also found https://bugs.openjdk.org/browse/JDK-8268854 and https://openjdk.org/jeps/8260688 which try to address the ReDoS issue in Java.

**Comment by merlimat** [comment]
It's to address the issues reported at  https://github.com/apache/pulsar/security/code-scanning/24

**Comment by heesung-sn** [comment]
LGTM, but I see regx errors from the tests.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java (hunk 1 above)
Comment (mattisonchao) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java:3043]: Yep, It should be a breaking changes here. :) 

```java
class Scratch {
    public static void main(String[] args) {
        final String regex = "a*b";
        final String quoteRegex = Pattern.quote(regex);
        final Pattern pattern1 = Pattern.compile(regex);
        final Pattern pattern2 = Pattern.compile(quoteRegex);
        final Matcher a = pattern1.matcher("aaab");
        System.out.println(a.find());
        final Matcher b = pattern2.matcher("aaab");
        System.out.println(b.find());
    }
}

```
true
false
```

**Comment by mattisonchao** [comment]
Hi guys, 
1. Maybe we can deprecate this method ` public static List<String> filterTopics(List<String> original, String regex)` to avoid alerts and then use ` public static List<String> filterTopics(List<String> original, Pattern topicsPattern)` instead. :) 
2. For the pattern-resolving problem, I've got two suggestions:
     1. Adopt https://github.com/google/re2j, as Lari provided above. 
     2. Do some defence programming to verify the time-consuming, and throw exceptions if the time is too long.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java (hunk 1 above)
Comment (merlimat) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java:3043]: Ouch. Let's close this then and focus on using re2 library

**Pull request closed without merging**